File: 11729.py
n = int(input())

def hanoi(r, start, end, step):
    if r == 1:
        print(f"{start} -> {end}")
    else:
        hanoi(r-1, start, step, end)
        print(f"{start} -> {end}")
        hanoi(r-1, step, end, start)

print(2**n -1)
if n <= 20:
    hanoi(n, 1, 3, 2)

File: 11729.cpp
#include <iostream>
#include <cmath>
using namespace std;

void hanoi(int n, int start, int step, int end) {
    if (n == 1) {
        cout << start << " " << end << "\n";
    }
    else {
        hanoi(n-1, start, end, step);
        cout << start << " " << end << "\n";
        hanoi(n-1, step, start, end);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int num;
    cin >> num;

    long long res = (long long)pow(2, num) - 1;
    cout << res << "\n";
    if (num <= 20) {
        hanoi(num, 1, 2, 3);
    }
    

    return 0;
}